java-checks/src/main/java/org/sonar/java/checks/synchronization/DoubleCheckedLockingCheck.java

```diff
@@ -0,0 +1,215 @@
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.BinaryExpressionTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.IfStatementTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.StatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+import javax.annotation.CheckForNull;
+import java.util.Collection;
+import java.util.Deque;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+
+import static org.sonar.plugins.java.api.tree.Tree.Kind.EQUAL_TO;
+import static org.sonar.plugins.java.api.tree.Tree.Kind.IDENTIFIER;
+import static org.sonar.plugins.java.api.tree.Tree.Kind.IF_STATEMENT;
+import static org.sonar.plugins.java.api.tree.Tree.Kind.MEMBER_SELECT;
+import static org.sonar.plugins.java.api.tree.Tree.Kind.NULL_LITERAL;
+import static org.sonar.plugins.java.api.tree.Tree.Kind.SYNCHRONIZED_STATEMENT;
+
+@Rule(key = "S2168")
+public class DoubleCheckedLockingCheck extends IssuableSubscriptionVisitor {
+
+  private Deque<CheckedLocking> ifConditionSymbolStack = new LinkedList<>();
+  private Deque<Tree> synchronizedStmtStack = new LinkedList<>();
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(IF_STATEMENT, SYNCHRONIZED_STATEMENT);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+    if (isIfFieldEqNullTree(tree)) {
+      IfStatementTree ifTree = (IfStatementTree) tree;
+      ifConditionSymbolStack.push(new CheckedLocking(ifTree));
+      visitIfStatement(ifTree);
+    }
+    if (tree.is(SYNCHRONIZED_STATEMENT)) {
+      synchronizedStmtStack.push(tree);
+    }
+  }
+
+  @Override
+  public void leaveNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+    if (isIfFieldEqNullTree(tree)) {
+      ifConditionSymbolStack.pop();
+    }
+    if (tree.is(SYNCHRONIZED_STATEMENT)) {
+      synchronizedStmtStack.pop();
+    }
+  }
+
+  private static boolean isIfFieldEqNullTree(Tree tree) {
+    if (!tree.is(IF_STATEMENT)) {
+      return false;
+    }
+    IfStatementTree ifTree = (IfStatementTree) tree;
+    if (!ifTree.condition().is(EQUAL_TO)) {
+      return false;
+    }
+    BinaryExpressionTree eqRelation = (BinaryExpressionTree) ifTree.condition();
+    return (isField(eqRelation.leftOperand()) && eqRelation.rightOperand().is(NULL_LITERAL))
+      || (isField(eqRelation.rightOperand()) && eqRelation.leftOperand().is(NULL_LITERAL));
+  }
+
+  private void visitIfStatement(IfStatementTree ifTree) {
+    if (insideCriticalSection()) {
+      CheckedLocking parentIf = sameFieldAlreadyOnStack(ifConditionSymbolStack.peek());
+      if (parentIf.ifTree == ifTree) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)